File: toolgui-wails/backend.go
package tgwails

import (
	"bytes"
	"context"
	"encoding/base64"
	"encoding/json"
	"sync"
	"uuid"

	"github.com/voilelab/toolgui/toolgui/tgframe"
	"github.com/voilelab/toolgui/toolgui/tgutil"

	"github.com/wailsapp/wails/v2/pkg/runtime"
)

// PackEventName is the Wails event every pack is emitted on. A single event
// name keeps create/update/delete/result in the order the page produced them.
const PackEventName = "toolgui:pack"

// ErrNoSession is returned when the frontend acts before calling Start.
var ErrNoSession = tgutil.NewError("no session, call Start first")

// ErrNoUpload is returned for an upload id the session doesn't know, which is
// what a chunk sent after the session was replaced looks like.
var ErrNoUpload = tgutil.NewError("no such upload")

// ToolGUI is the struct Wails binds. Its exported methods reach the frontend
// as window.go.tgwails.ToolGUI.<Method>, each returning a promise.
//
// Payloads cross as JSON strings, the same ones the web executor sends over
// its websocket, so both transports carry an identical wire format.
type ToolGUI struct {
	app *tgframe.App

	// emit hands one pack to the frontend. It's wired up when the window
	// starts, which is also what lets the tests run without one.
	emit func(packJSON string)

	// lock guards the session and the state it runs on: bound methods are
	// called from the frontend's goroutine, shutdown from another.
	lock    sync.Mutex
	session *tgframe.Session
	state   *tgframe.State

	// uploads are the files still arriving, by upload id. The lock guards
	// them too: a chunk that lands while the session is being replaced must
	// not write to the state that is going away.
	uploads map[string]*tgframe.File
}

// NewToolGUI return the bound struct serving app.
func NewToolGUI(app *tgframe.App) *ToolGUI {
	return &ToolGUI{app: app, uploads: make(map[string]*tgframe.File)}
}

// start is wired to [options.App.OnStartup] rather than being a bound method,
// so the frontend never sees it. The context it gets is what the Wails
// runtime needs to reach this window.
func (t *ToolGUI) start(ctx context.Context) {
	t.emit = func(packJSON string) {
		runtime.EventsEmit(ctx, PackEventName, packJSON)
	}
}

// shutdown is wired to [options.App.OnShutdown].
func (t *ToolGUI) shutdown(ctx context.Context) {
	t.lock.Lock()
	defer t.lock.Unlock()

	t.closeSession()
}

// AppConf return the app config as JSON. It's the desktop counterpart of
// GET /api/app.
func (t *ToolGUI) AppConf() (string, error) {
	bs, err := json.Marshal(t.app.AppConf())
	if err != nil {
		return "", tgutil.Errorf("%w", err)
	}

	return string(bs), nil
}

// Start open a session on pageName and run the page once. Calling it again
// switches page: the previous session is closed and the new one starts from
// an empty state, the same as loading another page in the browser.
func (t *ToolGUI) Start(pageName string) error {
	t.lock.Lock()

	t.closeSession()

	state := tgframe.NewState()
	session, err := tgframe.NewSession(t.app, pageName, state, t.send)
	if err != nil {
		t.lock.Unlock()
		return tgutil.Errorf("%w", err)
	}

	t.state = state
	t.session = session
	t.uploads = make(map[string]*tgframe.File)

	t.lock.Unlock()

	// Draw the page for the first time.
	session.HandleEvent(&tgframe.EventEmpty{})
	return nil
}

// Update apply a frontend event to the session and rerun the page. It's the
// desktop counterpart of the update websocket.
func (t *ToolGUI) Update(eventJSON string) error {
	session := t.currentSession()
	if session == nil {
		return ErrNoSession
	}

	err := session.HandleRawEvent([]byte(eventJSON))
	if err != nil {
		return tgutil.Errorf("%w", err)
	}

	return nil
}

// UploadFileStart opens a file for an upload and returns the id the chunks
// after it carry. It's the desktop half of POST /api/files, which the bridge
// can't do in one call: it takes strings, so a whole file would sit in memory
// as a blob, as base64 and as bytes at once.
func (t *ToolGUI) UploadFileStart(name string) (string, error) {
	t.lock.Lock()
	defer t.lock.Unlock()

	if t.state == nil {
		return "", ErrNoSession
	}

	file, err := t.state.NewFile(name)
	if err != nil {
		return "", tgutil.Errorf("%w", err)
	}

	uploadID := uuid.New().String()
	t.uploads[uploadID] = file

	return uploadID, nil
}

// UploadFileChunk appends one base64 encoded chunk to the upload.
func (t *ToolGUI) UploadFileChunk(uploadID, dataBase64 string) error {
	bs, err := base64.StdEncoding.DecodeString(dataBase64)
	if err != nil {
		return tgutil.Errorf("%w", err)
	}

	t.lock.Lock()
	defer t.lock.Unlock()

	file, ok := t.uploads[uploadID]
	if !ok {
		return ErrNoUpload
	}

	if err := file.Append(bytes.NewReader(bs)); err != nil {
		return tgutil.Errorf("%w", err)
	}

	return nil
}

// UploadFileFinish hands the finished upload to the component that asked for
// it. Until then the page doesn't see it, so a second pick on the same
// component replaces the first rather than being spliced into it.
func (t *ToolGUI) UploadFileFinish(componentID, uploadID string) error {
	t.lock.Lock()
	defer t.lock.Unlock()

	if t.state == nil {
		return ErrNoSession
	}

	file, ok := t.uploads[uploadID]
	if !ok {
		return ErrNoUpload
	}

	delete(t.uploads, uploadID)
	t.state.PutFile(componentID, file)

	return nil
}

// send push a pack to the frontend. [tgframe.Session] serializes the calls,
// and the Wails event queue preserves their order.
func (t *ToolGUI) send(pack any) error {
	if t.emit == nil {
		return tgutil.NewError("wails runtime is not ready")
	}

	bs, err := json.Marshal(pack)
	if err != nil {
		return tgutil.Errorf("%w", err)
	}

	t.emit(string(bs))
	return nil
}

func (t *ToolGUI) currentSession() *tgframe.Session {
	t.lock.Lock()
	defer t.lock.Unlock()

	return t.session
}

// closeSession must be called with lock held.
func (t *ToolGUI) closeSession() {
	if t.session == nil {
		return
	}

	t.session.Close()
	t.session = nil

	// The state owns the files uploaded to it, the ones still arriving
	// included, and nothing else can reach them once the session is gone.
	t.state.Destroy()
	t.state = nil
	t.uploads = make(map[string]*tgframe.File)
}

File: toolgui-web/wails/src/api/backend.ts
import { AppConf, DownloadResult, UpdateEvent, UploadResult } from "@toolgui-web/lib"

// Backend is the Go struct Wails binds. Every bound method returns a promise.
// Payloads cross as JSON strings, the same ones the websocket transport
// carries, so both lanes share a wire format.
interface Backend {
  AppConf(): Promise<string>
  Start(pageName: string): Promise<void>
  Update(eventJSON: string): Promise<void>
  UploadFileStart(name: string): Promise<string>
  UploadFileChunk(uploadID: string, dataBase64: string): Promise<void>
  UploadFileFinish(componentID: string, uploadID: string): Promise<void>
}

// WailsRuntime is the slice of window.runtime this adapter uses.
interface WailsRuntime {
  EventsOn(eventName: string, callback: (...data: any[]) => void): () => void
  BrowserOpenURL(url: string): void
}

declare global {
  interface Window {
    // Wails names bindings after the Go package and struct they came from.
    go: { tgwails: { ToolGUI: Backend } }
    runtime: WailsRuntime
  }
}

export function backend(): Backend {
  return window.go.tgwails.ToolGUI
}

export function onEvent(eventName: string, callback: (data: string) => void) {
  window.runtime.EventsOn(eventName, callback)
}

// openURL hands a url to the system's default browser. The window stays
// where it is, which is the point: see src/external_link.ts.
export function openURL(url: string) {
  window.runtime.BrowserOpenURL(url)
}

// getAppConf is the desktop counterpart of GET /api/app.
export async function getAppConf(): Promise<AppConf> {
  return JSON.parse(await backend().AppConf())
}

export function sendEvent(event: UpdateEvent): Promise<void> {
  return backend().Update(JSON.stringify(event))
}

// CHUNK_SIZE is how much of a file crosses the bridge at a time. Bindings
// take strings, so each chunk costs its own size again as base64.
const CHUNK_SIZE = 4 * 1024 * 1024

// uploadFile is the desktop counterpart of POST /api/files. It sends the file
// in chunks: reading it whole would hold it as a blob, as base64 and as bytes
// on the Go side all at once.
export async function uploadFile(file: File, componentID: string): Promise<UploadResult> {
  try {
    // The component gets the file only once every chunk has landed, so a
    // second pick while this one is still going replaces it rather than
    // being spliced into it.
    const uploadID = await backend().UploadFileStart(file.name)

    for (let offset = 0; offset < file.size; offset += CHUNK_SIZE) {
      const chunk = file.slice(offset, offset + CHUNK_SIZE)
      await backend().UploadFileChunk(uploadID, await toBase64(chunk))
    }

    await backend().UploadFileFinish(componentID, uploadID)
    return { ok: true }
  } catch (e) {
    return { ok: false, error: String(e) }
  }
}

// downloadFile is what the desktop build has instead of GET /api/files, which
// is nothing: there is no HTTP endpoint here and no origin private file system
// either, and how a bound method should hand a file of any size to the webview
// is not settled. So DownloadFile draws and fails to save, and the component
// is one the desktop build does not carry yet; DownloadButton works here, as
// it always has.
export async function downloadFile(_token: string): Promise<DownloadResult> {
  return {
    ok: false,
    error: 'the desktop build cannot fetch a download yet; use DownloadButton',
  }
}

// toBase64 drops the "data:<type>;base64," prefix FileReader adds.
function toBase64(file: Blob): Promise<string> {
  return new Promise((resolve, reject) => {
    const reader = new FileReader()
    reader.onload = () => {
      const result = reader.result as string
      resolve(result.slice(result.indexOf(',') + 1))
    }
    reader.onerror = () => reject(reader.error)
    reader.readAsDataURL(file)
  })
}
